### PR TITLE
Allow encrypted attachments to be forwarded

### LINKED ISSUE #3460: Attachement get lost by forwarding a (encrypted but not signed) Mail
### Expected behavior
If I forward a Mail, the whole mail including Attachments should be forwarded

### Actual behavior
The Content is correct in the forward mail but Attachments are missed

### Steps to reproduce
1. Receive a **encrypted but not signed** Mail with Attachment (e.g. test.pdf)
2. Press the Button to forward this Mail
3. Thats ist, the Attachment is missing

### Environment
K-9 Mail version: 5.502

Android version: 7.0

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

### PR DESCRIPTION
**Pull request by artbristol**
Fixes #3460

### CODE REVIEW

**Comment by artbristol** [comment]
@cketti I have tried to implement your comment from here https://github.com/k9mail/k-9/issues/3460#issuecomment-860209098

**Comment by cketti** [comment]
Looks like you unconditionally set `LOADER_ARG_INTERNAL` to `true`. This should only happen for attachments "added" by K-9 Mail (e.g. when forwarding a message or loading a draft). A good starting point for an investigation is `AttachmentPresenter.processMessageToForward()`.

**Comment by artbristol** [comment]
I still don't quite understand the issue unfortunately. 

Is it that an attacker can call K9's *content provider*, or that an attacker can create one of K9's *activities*? (Or both?)

`AttachmentPresenter` is K9's own code, only ever created from the `MessageCompose` activity and so I can't see danger from the content provider attack.

Is the problem that an attacker could create a MessageCompose activity (via Intent) and supply an attachment URI to be decrypted?

**Comment by cketti** [comment]
The content providers themselves are protected by Android's access control mechanisms (`android:exported="false"`, `android:grantUriPermissions="true"`). But this only affects direct access by external apps. 

The issue here is that K-9 Mail accesses the content providers using content URIs provided by external apps. So for the OS it looks like K-9 Mail is accessing its own content providers and of course that's allowed. But an external app could e.g. send a Share intent containing a content URI for `DecryptedFileProvider`. That external app can't access that content provider directly. But using the Share mechanism it could get K-9 Mail to attach the decrypted part as an attachment to a message that gets sent to an attacker-controlled email address (also supplied in the Share intent). 
This does require user interaction. But let's be honest, the vast majority of people won't be checking the contents of an attachment that is attached using a Share intent. The malicious app could e.g. pretend to send information about a crash to its developer (using K-9 Mail).

So, whenever a content URI is provided by an external app (Share intent, ACTION_GET_CONTENT), we need to use `SafeContentResolver`. When the content URI originates from K-9 Mail itself we can use `ContentResolver`.

**Comment by artbristol** [comment]
Re. verifying the Intent came from K-9: there doesn't seem to be an obvious way to do it. (Best I could find was https://dev.guardianproject.info/projects/trustedintents/wiki/Wiki/)

I've added a commit to verify that the intent was not ACTION_GET_CONTENT but it seems brittle and not future-proof.

Maybe K-9 could add a short-term secret to the Intent's extras, that wouldn't be known to external applications?

Or we could make the URI of an encrypted attachment cryptographically secure such that an external application wouldn't be able to guess it?

**Comment by cketti** [comment]
Luckily, the origin of the `Intent` is not important.

There are only a few ways how attachments can be added to a message:
- Starting `MessageCompose` using an `Intent` action of `com.fsck.k9.intent.action.*` and an extra of `EXTRA_MESSAGE_REFERENCE` or `EXTRA_MESSAGE_DECRYPTION_RESULT`. See `MessageCompose.processSourceMessage()`. K-9 Mail is using this internally. But the `Intent` extras only contain references to a message in the database, never `content:` URIs.
- Starting `MessageCompose` using `ACTION_SEND*` or `ACTION_VIEW`. The `Intent` will contain a `content:` URI. See `MessageCompose.initFromIntent()`. K-9 Mail never uses this internally.
- Using the "add attachment" action in the UI. See `MessageCompose.showPickAttachmentDialog()` and `AttachmentPresenter.onActivityResult()`. The result `Intent` will contain a `content:` URI. K-9 Mail never uses this internally with trusted `content:` URIs.

In the first case we want to use `ContentResolver`. In the other two cases `SafeContentResolver`.

**Comment by cketti** [comment]
Whether or not to trust a `content:` URI can't be a property of `AttachmentPresenter`. You can e.g. start `MessageCompose` by forwarding a message and then manually add an attachment. The attachments loaded from the source message are "internal" attachments, the manually added one is an "external" attachment. But there's only one `AttachmentPresenter` instance managing both operations.
The `Attachment` class already stores all kinds of information about an attachment including the loading state. I suggest we use it to store whether it's an internal or external attachment, too. `Attachment` instances also survive `Activity` recreations/process death (see `AttachmentPresenter.onRestoreInstanceState()`), which is exactly what we need.

To me it looks like you have a hard time figuring out where to start with this change. That's why I created the branch `attachment_source` as a starting point. https://github.com/k9mail/k-9/compare/attachment_source
I hope it can be helpful.

**Comment by artbristol** [comment]
Thanks. I did actually go down that path of adding the provenance to the attachment, but it all got too complicated. Let me know if the current branch seems better.

File: app/core/src/main/java/com/fsck/k9/message/Attachment.kt [hunk 1]
```diff
@@ -6,6 +6,7 @@ interface Attachment {
     val contentType: String?
     val name: String?
     val size: Long?
+    val internalAttachment: Boolean
```
Comment (cketti) [inline app/core/src/main/java/com/fsck/k9/message/Attachment.kt:9]: The API feels nicer if we change this to `isInternalAttachment`.

**Review by cketti — APPROVED** [review]
:+1: Looks good.

Can you please squash the commits.

File: app/core/src/main/java/com/fsck/k9/message/Attachment.kt (hunk 1 above)
Comment (artbristol) [reply app/core/src/main/java/com/fsck/k9/message/Attachment.kt]: Done. I didn't realise Kotlin took into account the whole JavaBean `boolean`/`Boolean` quagmire.

**Comment by cketti** [comment]
Thanks :+1:

**Pull request merged**